### PR TITLE
[ci2] add support for deploy

### PR DESCRIPTION
**Pull request by cseed**
- but don't enable
 - add back deployable to watched branches
 - add deploy flag to build configuration
 - added Code option that represents a version of code (but not quite a sha, because it might do a merge that produces a new sha) and can check itself out

I think this gets us pretty close.  Working on ci2 tests now.

### CODE REVIEW

File: ci2/ci/build.py [hunk 1]
```diff
@@ -230,19 +262,17 @@ def config(self):
...
                                           command=['bash', '-c', script],
                                           attributes={'name': self.name},
                                           volumes=volumes,
                                           input_files=input_files,
-                                          copy_service_account_name=sa,
                                           parent_ids=self.deps_parent_ids())
 
-    async def cleanup(self, batch, sink):
+    async def cleanup(self, batch, deploy, sink):
+        if deploy and self.publish_as:
```
Comment (danking) [inline ci2/ci/build.py]: if publish_as is none, then we never pushed the image so we don't need to cleanup, right?

File: ci2/ci/build.py [hunk 2]
```diff
@@ -316,60 +352,80 @@ def config(self):  # pylint: disable=no-self-use
...
             self.image,
             command=['bash', '-c', rendered_script],
             attributes={'name': self.name},
             input_files=input_files,
             output_files=output_files,
-            copy_service_account_name=sa,
-            parent_ids=self.deps_parent_ids())
+            volumes=volumes,
+            parent_ids=self.deps_parent_ids(),
+            always_run=self.always_run)
```
Comment (danking) [inline ci2/ci/build.py]: why is the RunImageStep always_run? If the image build fails, we shouldn't bother running it right?

File: ci2/ci/build.py [hunk 3]
```diff
@@ -448,7 +504,10 @@ def config(self):
                                           service_account_name='ci2-agent',
                                           parent_ids=self.deps_parent_ids())
```
Comment (danking) [inline ci2/ci/build.py]: Does this mean we're not going to include namespaces in the yaml files? Namespaces are always defined in the build configuration?

File: ci2/ci/github.py [hunk 4]
```diff
@@ -338,10 +377,28 @@ def __init__(self, branch):
         self.prs = new_prs
 
         for pr in new_prs.values():
-            await pr.refresh_review_state(gh)
+            await pr._refresh_review_state(gh)
+
+    async def _heal(self, batch_client):
+        if self.deployable and self.sha and not self.deploy_state:
+            if not self.deploy_batch:
+                deploy_branches = await batch_client.list_batches(
```
Comment (danking) [inline ci2/ci/github.py]: s/branches/batches/

File: ci2/ci/build.py (hunk 1 above)
Comment (cseed) [reply ci2/ci/build.py]: Cleanup deletes temporary images (pushed to `gcr.io/<project>/ci-intermediate:<token>`)

File: ci2/ci/build.py (hunk 2 above)
Comment (cseed) [reply ci2/ci/build.py]: `always_run` is for cleanup steps, namely, they run whether the parents (or ancestors) run successfully or not.  So we set `always_run` explicitly on for ci to clean up after itself (e.g. delete temporary images, files, namespaces, etc.)

The user also might want to cleanup in the testing pipeline. For example, testing ci, we want to create a temporary github repo that should be cleaned up after the tests.

Skipping if the image fails is reasonable, although I don't think it hurts to run the job and have it fail, and it isn't easy to do this in batch right now.

File: ci2/ci/build.py (hunk 3 above)
Comment (cseed) [reply ci2/ci/build.py]: I'm not quite sure which code you're referring to.

Correct.  Namespaces should not be able to be referred to directly.  runImage is run by batch, so executes in batch-pods.  deploy takes yaml, but requires a namespace which I verify is was created by a `createNamespace` step, see `get_namespace`.  My longer term intention is to lock down the yaml files by whitelisting k8s features and parsing/validating them.  Right now, I explicitly create them with `kubectl -n <namespace>` which restricts itself to a namespace according to the documentation (although I don't know exactly what that means).

Does that answer your question?

**Pull request merged**